contrib/datadog/tracing/interceptor.go

```diff
@@ -0,0 +1,262 @@
+// The MIT License
+//
+// Copyright (c) 2021 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package tracing provides DataDog tracing utilities
+package tracing
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"strconv"
+	"strings"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"go.temporal.io/sdk/interceptor"
+	"go.temporal.io/sdk/log"
+	"go.temporal.io/sdk/workflow"
+)
+
+// InterceptorOptions are options provided to NewInterceptor
+type InterceptorOptions struct {
+	// DisableSignalTracing can be set to disable signal tracing.
+	DisableSignalTracing bool
+
+	// DisableQueryTracing can be set to disable query tracing.
+	DisableQueryTracing bool
+
+	// DisableLogTraceLinking can be set to disable the automatic addition of "dd.trace_id" and "dd.span_id" fields to the logger
+	// provided by this interceptor.
+	DisableLogTraceLinking bool
+
+	// SpanContextKey is the context key used for internal span tracking. If not
+	// set, this defaults to an internal key (recommended).
+	SpanContextKey interface{}
+
+	// HeaderKey is the Temporal header field key used to serialize spans. If
+	// empty, this defaults to the one used by all SDKs (recommended).
+	HeaderKey string
```

Comment:
What is the use case for making these configurable? I think we can always add higher level fields to extract/inject spans from carriers if necessary.

I would prefer not to include these options, since this makes it difficult/impossible to ensure tracing compatibility between language SDKs or even differently configured worker processes.

Reply:
Discussed with @cretz on Slack, and we're good to remove these options for now.

The use case was for handling custom spans in workflow code. I would like to tackle this with a higher level API that generates deterministic span IDs to make it safe to trace long-running pieces of workflow code.

Reply:
I believe the use case was to support multiple tracing implementations at once. For example if you wanted to have OpenTracing and OpenTelemetry, you'd need to disambiguate the headers. But I have no problem with y'all removing this from these options.